ボウリングのスコア計算プログラムを作成しました

ksmxxxxxx opened the pull request:
# 概要

ボウリングのスコア計算プログラムを作成しました。

# やったこと

- スコア計算プログラムの作成
- プログラムのクラスの計算メソッド毎のテスト
  - テストパターンは[プラクティスページ](https://bootcamp.fjord.jp/practices/210)の例をテストするようにしています
- ターミナルでの実行結果をスクリーンショット

# 特にレビューしてほしいと思っているポイント

- メソッドの分割
  - RubocopのLint結果を参考に、なるべく意味のある単位になるように分割した（が、これで方向性としてあってるのか？）
- Gameクラスの処理
  - 条件分岐がかなり複雑になってしまったのが気になっています。
  - もしかしたら`add_zero_when_strike`メソッドの処理しなくてもいいのでは？とひととおり作ってから考えたのですが、これをしないと10フレーム目の計算とかが合わなかったとか……何か理由があってこの処理を入れているんですが、作った当初から時間が経ちすぎてて入れた理由が思い出せず…。

# Screenshot (Terminal result)
<img width="303" alt="スクリーンショット 2021-01-19 11 03 53" src="https://user-images.githubusercontent.com/5976902/104979091-c8eb3380-5a46-11eb-86de-8d9bda5f8951.png">

ksmxxxxxx committed:
:tada: Create bowling score calc program with object oriented programming

torinoko commented on 07.bowling_object/lib/bowling.rb:
```diff
@@ -0,0 +1,126 @@
+
+  def score
+    frames = divide_frame
+    frames.size
```
ここの行はどんな意図で入れていますか？

ksmxxxxxx replied:
多分作り途中の確認用コードの名残ですね……削除します。

ksmxxxxxx replied:
🚮 https://github.com/ksmxxxxxx/ruby-practices/pull/1/commits/11ce4643f14817497342510498572991f7ee840c

torinoko commented on 07.bowling_object/lib/bowling.rb:
```diff
@@ -0,0 +1,126 @@
+    frames = divide_frame
+    frames.size
+    10.times.each do |num|
+      add_frame_score(frames, num)
```
メソッド名が `add_frame_score` で `frame` が単数系ですが、渡しているのは `flames` で複数なんですね。命名の分かりやすさはもう少し上げることができそうな気がします。

ksmxxxxxx replied:
Fixed
a5861e01fcde996ed3c76e592e2039a9d4ba03f1

torinoko commented on 07.bowling_object/lib/bowling.rb:
```diff
@@ -0,0 +1,126 @@
+    10.times.each do |num|
+      add_frame_score(frames, num)
+    end
+    all_score.sum
```
`@all_score` は最終的にここで合計するんですね。
`@all_score` をインスタンス変数にするメリットってどんなところでしょう？

ksmxxxxxx replied:
Fixed
42e9b189a0de2b1bbbbe27ede4dd816aea48ee34
201b908a38c01cec5c20315c3c034992f3d61a80

メソッド間で配列を渡すためにインスタンス変数にしていたんですが、2つのメソッドでしか使ってなかったし、 `score`メソッドの中で計算するために配列に入れていってるだけなので、`score`メソッドの中のローカル変数に `[]`を作って上げて、そこにどんどこ追加して`sum`すれば良いかと思ってこの変更をしました。

指摘意図、あってたでしょうか？

torinoko commented on 07.bowling_object/lib/bowling.rb:
```diff
@@ -0,0 +1,126 @@
+  end
+
+  def add_frame_score(frames, num)
+    all_score << if num <= 8
+                   calc_frame_first_between_nineth(frames, num)
+                 else
+                   calc_frame_last(frames, num)
+                 end
```
これは好みの問題なのですが、`if` の戻り値を変数に代入するような書き方はあまりしないです。
一般的ではないので読み手がちょっとがんばらなくちゃいけないかな？

torinoko replied:
引数の `frames` と `num` ですが、メソッドの中でまた次のメソッドに渡され続けているのがちょっと気になりますね。
特に `frames` はインスタンス変数にもなので、全体的にもう少しすっきりさせることができそうな気がします。

ksmxxxxxx replied:
> これは好みの問題なのですが、if の戻り値を変数に代入するような書き方はあまりしないです。
一般的ではないので読み手がちょっとがんばらなくちゃいけないかな？

https://www.rubydoc.info/github/bbatsov/RuboCop/RuboCop/Cop/Style/ConditionalAssignment

こちらはRubocopのLintに合わせて書いたものになります。
デフォルトの設定だと警告が出るパターンだったのですが、fjordのrubocopは警告出さないようになってるんですかね？
（提出した時rubocopは通してたんですが、先日rubyにインストールされてるrubocopがfjordのものじゃなかった可能性があることが一昨夜発覚しました…）

ksmxxxxxx replied:
あー……っていうより、if分の部分をメソッドとかにしてしまって、変数にメソッドを代入するって意図ですかね？

torinoko replied:
そうですね、処理が短ければ三項演算子で、長くなりそうであればメソッドにするなどの対応をとることが多いですね！

ksmxxxxxx replied:
Fixed
https://github.com/ksmxxxxxx/ruby-practices/pull/1/commits/f2768285a771f24897602830b27ecd34018ebbf9
1839a53d6ff1c3e3a59fec24e2a382356d014570

torinoko commented on 07.bowling_object/lib/bowling.rb:
```diff
@@ -0,0 +1,126 @@
+      Frame.new(frames[num][0], frames[num][1], frames[num + 1][0]).score
+    else
+      Frame.new(frames[num][0], frames[num][1], 0).score
+    end
```
`if` の中に `if` がある（ `if` がネストしているとも言いますね :)）のは、コードをよくできる可能性のあるところです。
メソッドに切り出せたり、処理をコンパクトにできるかも？

`Frame.new` がたくさん登場しているのも、重複している処理があるということなので改善チャンスがありそうです。

ぜひ考えてみてください〜！

ksmxxxxxx replied:
Fixed
3659f4eb6c037951d9b5f3803181b055e1c2fb97

意図が組めてるかわかりませんが、こういうことかなという修正をしてみました。

torinoko commented on 07.bowling_object/lib/bowling.rb:
```diff
@@ -0,0 +1,126 @@
+      end
+    else
+      Frame.new(frames[num][0], frames[num][1], frames[num + 1][0]).score
+    end
```
ここもそうですね！

ksmxxxxxx replied:
Fixed
3659f4eb6c037951d9b5f3803181b055e1c2fb97

torinoko commented on 07.bowling_object/lib/bowling.rb:
```diff
@@ -0,0 +1,126 @@
+
+  def divide_frame
+    add_zero_when_strike.each_slice(2).map { |s| s }
+  end
```
ここでは具体的にどんな処理をしているんでしょうか？
`map` メソッドについても見直してみるといいかもしれません。

ksmxxxxxx replied:
作ったフレームを1フレーム2投にするための処理だったんですが、必要なかったのでこの処理はなくしました。
8096d8e

torinoko commented on 07.bowling_object/lib/bowling.rb:
```diff
@@ -0,0 +1,126 @@
+    add_zero_when_strike.each_slice(2).map { |s| s }
+  end
+
+  def add_zero_when_strike
+    game_score.each do |shot|
+      if shot.include?('X')
+        shots << 'X'
+        shots << '0' if shots.size.odd?
+      else
+        shots << shot
+      end
+    end
+    shots
+  end
```
`add_zero_when_strike` というメソッド名に比べるといろんな処理が入っていますね。
最後にインスタンス変数の `@shots`を返していますが、ここでしか使ってないかな？
インスタンス変数と、メソッド内でのみ使う変数を分けて使えるとよさそうですね。

ksmxxxxxx replied:
Fixed
ac96fc43d59dd47b7a2758cfd53dc5e77eaa275e

ksmxxxxxx replied:
`X`の時、`0`を挿入するものでしたが、そもそも入れなくても計算できるなと思ったので、変更しました。
8096d8e

torinoko commented on 07.bowling_object/lib/bowling.rb:
```diff
@@ -0,0 +1,126 @@
+    first_shot.score == 10
+  end
+
+  def spare?
+    [first_shot.score, second_shot.score].sum == 10
+  end
```
スペアは「1投目で残ったピンを2投目で全部倒すと10点」なんですね〜。
これだとストライクも `spare?` で `true` を返しちゃいますね。
ストライクのときは `false` を返せるとよさそうですね。

ksmxxxxxx replied:
Fixed
6661a9a

torinoko commented on 07.bowling_object/lib/bowling.rb:
```diff
@@ -0,0 +1,126 @@
+  end
+
+  def score
+    return 10 if mark.include?('X')
+
+    mark.to_i
```
このくらいであれば三項演算子を使ってもいいかもしれませんね。
（好みの問題ですし、三項演算子を嫌う人もいるのでどちらでも〜！）
```
mark.include?('X') ? 10 : mark.to_i
```

ksmxxxxxx replied:
Fixed
https://github.com/ksmxxxxxx/ruby-practices/pull/1/commits/f2768285a771f24897602830b27ecd34018ebbf9

torinoko reviewed (changes requested):
気になったところをコメントしました。
コードも増えてきましたし、クラスごとにファイルを分けてみてもいいかもですね。

> メソッドの分割
> RubocopのLint結果を参考に、なるべく意味のある単位になるように分割した（が、これで方向性としてあってるのか？）

「なるべく意味のある単位」を考慮したのはとてもよいですね！
方向性は大丈夫だと思います。

> Gameクラスの処理
> 条件分岐がかなり複雑になってしまったのが気になっています。

「複雑さが気になる」という感覚はすごくいいですね。
私も気になってます！
個別のコメントにも書きましたが、変数や処理で見直せそうなポイントがありますね。

> 何か理由があってこの処理を入れているんですが

お、ここの理由ははっきりさせたい気がしますね！
はっきりしない処理のせいで、複雑さが増している可能性もあります。
あと、コードはこの後メンテナンスされるかもしれないですよね。
そのとき毎回「これ何だっけ？」って考えるのは面倒ですよね。
面倒なことは今のうちに片付けちゃいましょう！

ksmxxxxxx committed:
:put_litter_in_its_place: Needless code

ksmxxxxxx committed:
:pencil2: Rename to add_frames_score

ksmxxxxxx committed:
:sparkles: Simplify method arguments

ksmxxxxxx committed:
:sparkles: Simplify if statements

ksmxxxxxx committed:
:recycle: Change @all_score to local variable

Did not have to use instance variable

ksmxxxxxx committed:
:sparkles: Simplify if statement and correct duplication

- Methodization nested if statement
- Methodization corrected duplication instance

ksmxxxxxx committed:
:recycle: Change variable and Rename method name

- Did not have to use instance variable
- The meaning of the name did not match the proccessing content

ksmxxxxxx committed:
:recycle: Change return false when strike

ksmxxxxxx committed:
:sparkles: Simplify methods of calc game total score

ksmxxxxxx committed:
:recycle: Split Files

ksmxxxxxx commented:
> コードも増えてきましたし、クラスごとにファイルを分けてみてもいいかもですね。

2639cb4

ここでファイルを分割しました。

torinoko commented on 07.bowling_object/lib/frame.rb:
```diff
@@ -0,0 +1,32 @@
+    first_shot.score == 10
+  end
+
+  def spare?
+    [first_shot.score, second_shot.score].sum == 10 unless strike?
+  end
```
`strike?` が true のときに何が返るでしょうか？
`?` のついてるメソッドなので、true / false が返るとよいですね 😃

ksmxxxxxx replied:
あー、これは私のほうがちょっと勘違いしてました 💦 
`[first_shot.score, second_shot.score].sum == 10`だけだと、`strike?`も通っちゃうから、
`first_shot.score`が`strike?`でfales、かつ`[first_shot.score, second_shot.score].sum == 10`じゃないとspareにならないですね
修正します〜

ksmxxxxxx replied:
4c4a8f5

こういうことかな…と考えて修正してみました。
が、ちょっとゲシュタルト崩壊気味です……🤪

torinoko replied:
そうですそうです、完璧ですー！
元のコードだと、 `strike?` が `true` だったときに `nil` が返っちゃうんですよね。

意味のわからないコードで申し訳ないのですが、
下記のようなメソッドを irb 等で実行してみてください。
引数でゼロが渡されたときに `nil` が返ることを確認できると思います。
```
def hoge?(num)
  true unless num.zero?
end
```

なので修正いただいたように、条件の中に `!strike?` を入れてしまうとこれを回避できます。

torinoko commented on 07.bowling_object/lib/game.rb:
```diff
@@ -0,0 +1,72 @@
+    end
+  end
+
+  def frame_is_last
+    if game_frames[num].size == 2
+      Frame.new(game_frames[num][0], game_frames[num][1], '3').score
+    else
+      Frame.new(game_frames[num][0], game_frames[num][1], game_frames[num][2]).score
+    end
+  end
```
ここはこんなふうに書き換えることもできそうですね。
（うーんでも分かりにくくなっちゃうかな……どうだろ……）
```
def frame_is_last
  third_mark = game_frames[num].size == 2 ? '3' : game_frames[num][2]
  Frame.new(game_frames[num][0], game_frames[num][1], third_mark).score
end
```

ksmxxxxxx replied:
これはちょっと迷いますね。
@torinoko さんの案もスッキリしてていいかなぁと思う一方で、パット見が分かりづらいかなぁという気がします。

ちなみに
```
Frame.new(game_frames[num][0], game_frames[num][1], '3').score
```
の最後の`'3'`は、デバック用に適当に3を入れてただけなので、0にしようとし思います。

ksmxxxxxx replied:
7cd8cda

第3引数の数字だけなおして、if文のところはそのままにしました。
pushしたあとに

```
  def frame_is_last
    first_mark = game_frames[num][0]
    second_mark = game_frames[num][1]
    third_mark = game_frames[num].size == 2 ? '0' : game_frames[num][2]
    Frame.new(first_mark, second_mark, third_mark)
  end
```
こういうのだったらもうちょっと見やすいかなぁって考えたんですがどうでしょう？
冗長ですかね？

torinoko replied:
好みはあると思いますが、私は好きです！
一時的にしか使わないとしても、意味のある名前がついた変数に入れるとすごく読みやすくなりますねー。

ksmxxxxxx replied:
ありがとうございます！この方向で修正しておきます〜 👏

torinoko reviewed (approved):
@ksmxxxxxx 
よいと思います！
見通しのよいコードになりましたね ✨ 
`spare?` だけ修正必要そうですが、全体的にはよさそうなので Approve しますね！

ksmxxxxxx committed:
:bug: Fix debug argument

ksmxxxxxx committed:
:bug: Fix spare method

ksmxxxxxx commented:
`git rebase -i` ed.
